### DISCUSSION TITLE
CoAP instead of MQTT

### DISCUSSION THREAD

**Discussion by GatCode** [arendst/Tasmota#14015] [Ideas]
Hello, is there a way to use Tasmota with CoAP instead of MQTT? This would be way more lightweight (UDP, not always connected,…). If not, is this a feature that might be integrated in the future? Thanks!

**Comment by sfromis** [comment]
MQTT is already really lightweight, having an an open connection is not "heavy", not the extra flows needed for HTTP. In general, chances of big changes depend a lot on "someone" willing to do the work and submit a PR with a feature fitting well with the code base. Much of what Tasmota does is event-based, a better fit for MQTT than for CoAP.

**Comment by GatCode** [comment]
Yes, it‘s lightweight but not as much as CoAP. An open connection is not ideal if you think about battery operated nodes. Also why is it necessary to use TCP - UDP is good enough. Don‘t get me wrong, I like MQTT but I think Tasmota could benefit from CoAP, especially since CoAP also supports resource observing.

**Reply by barbudor** [reply]
One of the great advantage of MQTT is that it's quite widespread among home automation users. It's a standard feature of many HA software (Home Assistant, OpenHab, Iobroker, Node-red,....)
I'm may be biased of course but i don't I've heard about CoAP as much as we hear about MQTT.

But as sfromis said, if it becomes an unavoidable protocol in that domain, why not if a PR can fit in the current base code (which I must say is probably a challenge seeing how tas**M**ota and MQTT are tied together

**Reply by sfromis** [reply]
When it comes to battery nodes, this is not much of an argument due to Tasmota already not running on a MCU suitable for battery devices. I just do not see much of a benefit from CoAP for where Tas**M**ota currently can function. Not meant as an objection, more about why the work would be worth it.

**Reply by s-hadinger** [reply]
We might add UDP to Berry in the future (ESP32 only). When this is done, it should be easy to send and receive CoAP.

**Reply by GatCode** [reply]
Thanks to all of you! I don‘t know why CoAP hasn’t become as widely adopted as MQTT but I guess it‘s because MQTT is really old and CoAP somewhat new. Technically CoAP is the better option and it also removes the Broker from the equation and therefore another point of failure. Nevertheless, I am very curious what the future will bring to Tasmota!

**Reply by barbudor** [reply]
How do you remove the broker from the equation ? 
Unless it is using Multicast which won't go through many networks.
UDP brings another question regarding QoS. As MQTT is using MQTT, normally messages are delivered (it could be at the expense of retransmission) while UDP is officially not guaranteed to be forwarded 100%

**Reply by sfromis** [reply]
From a technical standpoint, I prefer the structure of MQTT instead of CoAP, having the broker is one argument in *favor*. Where I see CoAP mostly make sense is for battery devices, but that is not a good fit for Tasmota anyway.

The future brings nothing, someone may decide that the work to bring something to Tasmota is worth it, if there is a good and clear case to be made.

**Reply by arturohernandez10** [reply]
I think it is a good candidate for battery devices. The use of broker less UDP would make it a cinch. 
 
* TCP has errors (timeouts) just like any other communication. All applications have to consider communication failures.
* There are MANY reasons why node to node communication can and does fail. Including MQTT.
* The level of reliability needs to be considered at the application level. It's simple the send a UDP ACK back to the sender, or else retransmit. Which is in essence what TCP does anyway.
* For small packets. UDP with IP multicast almost does what MQTT is offering, to home users, without the overhead. 
* For new potential users. Buying, configuring and maintaining an extra raspberry pi, even if they are IT personnel. Is an added complication. Flashing a few devices, is significantly less work. Most likely there are many who already have a Pi. The question is, how inclusive does Tasmota want to be?

**Reply by ascillato** [reply]
> The question is, how inclusive does Tasmota want to be?

Hi, I understand the hook you want to make, but as Tasmota is not a sponsored project, the real question is Who wants to code it and share it?
Tasmota is open to new Pull Requests and new features. The search you have to make is to find someone willing to spend time working on this feature and sharing it.

Tasmota project has a code of conduct that enforces inclusion and forbids discrimination.

**Reply by ascillato** [reply]
> For new potential users. Buying, configuring and maintaining an extra raspberry pi, even if they are IT personnel. Is an added complication.

Yes, but you can use a cloud MQTT server if you want without the need of extra hardware. The extra hardware is for maintaining all your information locally.

**Reply by sfromis** [reply]
Since the Berry language available with Tasmota on ESP32 has recently been enhanced with [support for UDP](https://tasmota.github.io/docs/Berry/#udp-class), this should make it quite a bit easier for someone convinced of CoAP benefits to start playing around with making it work for them.

**Reply by dooblem** [reply]
The MQTT broker is a significant Single Point Of Failure. This is wat I don't like about MQTT.
I'm looking for ideas for tasmota to communicate with a shelly, in a reliable way. The shelly includes CoAP (a variant named coiot).
Using MQTT is not an option for me as I don't want to depend on another PC or rpi.
I will probably fall back on using HTTP, even it's an heavy protocol.

**Reply by Noschvie** [reply]
`The MQTT broker is a significant Single Point Of Failure. This is wat I don't like about MQTT.`

why not running the MQTT Broker in a Kubernetes Cluster ?
I know it's a overkill ... :-)

**Comment by s-hadinger** [comment]
There are no plans to support CoAP. You are free to propose a PR to support this protocol. Although I'm not seeing a wide adoption of CoAP within the hobbyist community.

**Comment by wjohn007** [comment]
@dooblem 

I developed the [Berry-UdpBroker](https://github.com/wjohn007/Berry-UdpBroker) for the reasons you already mentioned.
Maybe this is a solution or at least an inspiration.

**Reply by dooblem** [reply]
Hello @wjohn007 ,
Yes it could be. Unfortunately the device I'm working on currently is an esp8266. My understanding is that its less powerfull than the esp32, so Tasmota comes without Berry support with it.
Anyway I will probably fall back using plain HTTP requests.
